### PR TITLE
Attempt to fix sourcebuild issues

### PR DESCRIPTION
**Pull request by wtgodbe**
Attempt to fix https://github.com/dotnet/source-build/issues/3925, https://github.com/dotnet/source-build/issues/3924

### CODE REVIEW

**Review by MichaelSimons — APPROVED** [review]

File: src/ubuntu/22.04/cross/arm64/Dockerfile [hunk 1]
```diff
@@ -4,13 +4,13 @@ FROM mcr.microsoft.com/dotnet-buildtools/prereqs:ubuntu-22.04-crossdeps-local
...
-ENV NODE_VERSION 20.9.0
+# install node
 RUN cd ~ && \
-    curl -LO https://nodejs.org/dist/v${NODE_VERSION}/node-v${NODE_VERSION}-linux-arm64.tar.xz && \
-    tar xf node-v${NODE_VERSION}-linux-arm64.tar.xz
\ No newline at end of file
+    curl -sL https://deb.nodesource.com/setup_20.x -o nodesource_setup.sh && \
+    sudo bash nodesource_setup.sh && \
+    sudo apt-get install nodejs -y && \
+    rm -rf /var/lib/apt/lists/*
```
Comment (ViktorHofer) [inline src/ubuntu/22.04/cross/arm64/Dockerfile:16]: Should we also rm the `nodesource_setup.sh` or doesn't it matter?

**Review by ViktorHofer — APPROVED** [review]
Out of curiosity, what's the node&npm version on Fedora 39? And I assume you tested both changes?

File: src/fedora/39/amd64/Dockerfile [hunk 2]
```diff
@@ -66,4 +66,5 @@ RUN dnf --setopt=install_weak_deps=False install -y \
 # Dependencies for Aspnetcore
 RUN cd ~ && \ 
     curl -sL https://rpm.nodesource.com/setup_20.x
-RUN dnf install -y nodejs && dnf clean all
\ No newline at end of file
+RUN dnf install -y nodejs && dnf clean all
+ENV NO_UPDATE_NOTIFIER
```
Comment (mthalman) [inline src/fedora/39/amd64/Dockerfile:70]: I think this needs to be set to a value.
```suggestion
ENV NO_UPDATE_NOTIFIER=true
```

Comment (ViktorHofer) [reply src/fedora/39/amd64/Dockerfile]: According to the docs it doesn't need to be set to any value but it doesn't harm: https://www.npmjs.com/package/update-notifier#user-settings

**Comment by wtgodbe** [comment]
> Out of curiosity, what's the node&npm version on Fedora 39? And I assume you tested both changes?

I don't work with Docker much, so normally I use the images that get produced by the build of this repo - how can I test them out from the sources?

File: src/ubuntu/22.04/cross/arm64/Dockerfile (hunk 1 above)
Comment (wtgodbe) [reply src/ubuntu/22.04/cross/arm64/Dockerfile:16]: @mthalman what do you think?

Comment (mthalman) [reply src/ubuntu/22.04/cross/arm64/Dockerfile:22]: Might as well. It won't be needed in the final image. I'm a fan of image cleanliness.

**Comment by mthalman** [comment]
> > Out of curiosity, what's the node&npm version on Fedora 39? And I assume you tested both changes?
> 
> I don't work with Docker much, so normally I use the images that get produced by the build of this repo - how can I test them out from the sources?

[Instructions for building](https://github.com/dotnet/dotnet-buildtools-prereqs-docker?tab=readme-ov-file#how-to-modify-or-create-a-new-image)

**Comment by mthalman** [comment]
```
#9 7.066 Unpacking nodejs (20.11.0-1nodesource1) ...
#9 8.642 dpkg: error processing archive /var/cache/apt/archives/nodejs_20.11.0-1nodesource1_amd64.deb (--unpack):
#9 8.642  trying to overwrite '/usr/include/node/common.gypi', which is also in package libnode-dev 12.22.9~dfsg-1ubuntu3
#9 8.662 dpkg-deb: error: paste subprocess was killed by signal (Broken pipe)
#9 8.671 Errors were encountered while processing:
#9 8.671  /var/cache/apt/archives/nodejs_20.11.0-1nodesource1_amd64.deb
#9 8.690 E: Sub-process /usr/bin/dpkg returned an error code (1)
```

**Comment by wtgodbe** [comment]
That's the same error as https://github.com/dotnet/dotnet-buildtools-prereqs-docker/pull/957#issuecomment-1894307239, @ViktorHofer it looks like the node uninstall isn't working, or is incomplete?

File: src/fedora/39/amd64/Dockerfile [hunk 3]
```diff
@@ -66,4 +66,5 @@ RUN dnf --setopt=install_weak_deps=False install -y \
 # Dependencies for Aspnetcore
 RUN cd ~ && \ 
     curl -sL https://rpm.nodesource.com/setup_20.x
-RUN dnf install -y nodejs && dnf clean all
\ No newline at end of file
+RUN dnf install -y nodejs && dnf clean all
+ENV NO_UPDATE_NOTIFIER=true
```
Comment (mthalman) [inline src/fedora/39/amd64/Dockerfile:70]: This should be applied consistently across all the Node-related Dockerfiles.

**Comment by ViktorHofer** [comment]
The uninstall worked for me locally on my Ubuntu 22.04 machine. Unsure what's required for this dockerfile.

**Comment by wtgodbe** [comment]
Alright, I think this is working now & feedback has been addressed.

File: src/fedora/39/amd64/Dockerfile (hunk 3 above)
Comment (wtgodbe) [reply src/fedora/39/amd64/Dockerfile:70]: Done

File: src/ubuntu/22.04/cross/arm64/Dockerfile (hunk 1 above)
Comment (wtgodbe) [reply src/ubuntu/22.04/cross/arm64/Dockerfile]: Done

**Review by mthalman — APPROVED** [review]

File: src/ubuntu/22.04/cross/arm64/Dockerfile
Comment (mthalman) [inline src/ubuntu/22.04/cross/arm64/Dockerfile:1]: @wtgodbe - Was it intentional to apply this to the crossbuild Dockerfile? This isn't used by source build.

Comment (wtgodbe) [reply src/ubuntu/22.04/cross/arm64/Dockerfile:1]: Oh, I assumed this file corresponded to `ubuntu-22.04-arm64`, but I guess https://github.com/dotnet/dotnet-buildtools-prereqs-docker/blob/main/src/ubuntu/22.04/Dockerfile covers both the regular ubuntu-22.04 image & the arm64 one. Should we revert these changes then?

Comment (mthalman) [reply src/ubuntu/22.04/cross/arm64/Dockerfile:1]: Yes, the change should be reverted and applied to the other Dockerfile. I can take care of it. I just wanted to confirm.

**Pull request merged**